### PR TITLE
Update to new specifications for proof-of-work challenges

### PR DESCRIPTION
**Pull request by proton-jsadun**

### CODE REVIEW

File: challenge.go [hunk 1]
```diff
@@ -2,43 +2,94 @@ package srp
...
+func Argon2PreimageChallenge(b64Challenge string) (b64Solution string, err error) {
 	challenge, err := base64.StdEncoding.DecodeString(b64Challenge)
 	if err != nil {
-		return 0, err
+		return "", err
+	}
+
+	// Argon2 challenges consist of 3 PRF keys, the hash output, and 4 32-bit argon2 parameters
+	if len(challenge) !=  3 * argon2PRFKeySize + sha256.Size + 4 * 4 {
+		return "", errors.New("srp:invalid Argon2 preimage challenge length")
```
Comment (wussler) [inline challenge.go:63]:
```suggestion
		return "", errors.New("srp: invalid Argon2 preimage challenge length")
```

**Comment by marinthiercelin** [comment]
Please update the changelog before merging

File: Changelog.md [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 ## Changed
 * Update `github.com/cronokirby/saferith` dependency to v0.33.0. Adds assembly routines support for more platforms
+* Update `ECDLPChallenge` to the new specification and add the new `Argon2PreimageChallenge`.
```
Comment (wussler) [inline Changelog.md:7]:
```suggestion
## Added
* Helpers for new challenge type `Argon2PreimageChallenge`.

## Changed
* Update `github.com/cronokirby/saferith` dependency to v0.33.0. Adds assembly routines support for more platforms
* Update `ECDLPChallenge` to the new specification.
```

File: challenge.go [hunk 3]
```diff
@@ -2,43 +2,107 @@ package srp
...
 	"golang.org/x/crypto/curve25519"
 )
 
-const ECDLPEphemeralSize = 24
+const ecdlpPRFKeySize = 32
 
-func ECDLPChallenge(b64Challenge string) (solution int64, err error) {
-	if len(b64Challenge) != 76 { // 56 bytes in base64
-		return 0, errors.New("srp:invalid ECDLP challenge length")
+func ECDLPChallenge(b64Challenge string, ctx context.Context) (b64Solution string, err error) {
```
Comment (wussler) [inline challenge.go:18]: I highly doubt this is going to work with go-mobile. Can we use a simpler solution like an int with seconds?

File: challenge.go [hunk 4]
```diff
@@ -2,43 +2,115 @@ package srp
...
+type deadlineExceededError struct{}
+
+func (deadlineExceededError) Error() string   { return "srp: deadline exceeded calculating proof-of-work challenge" }
+func (deadlineExceededError) Timeout() bool   { return true }
+func (deadlineExceededError) Temporary() bool { return true }
+
+const ecdlpPRFKeySize = 32
+
+// Deadlines are measured on the wall clock, not the monotonic clock, due
+// to unreliability on mobile devices
```
Comment (wussler) [inline challenge.go:29]: Top-level comments should be in the form `functionName` .... and terminated by a dot
```suggestion
// ECDLPChallenge computes the base64 solution for a given ECDLP base64 challenge 
// within deadlineUnixMilli milliseconds, if any was found. Deadlines are measured on the
// wall clock, not the monotonic clock, due to unreliability on mobile devices.
```

File: challenge.go [hunk 5]
```diff
@@ -2,43 +2,115 @@ package srp
...
+	solution := []byte{}
+	solution = append(solution, buffer...)
+	solution = append(solution, point...)
+
+	return base64.StdEncoding.EncodeToString(solution), nil
+}
 
+const argon2PRFKeySize = 32
+
+func Argon2PreimageChallenge(b64Challenge string, deadlineUnixMilli uint64) (b64Solution string, err error) {
```
Comment (wussler) [inline challenge.go:72]:
```suggestion
// Argon2PreimageChallenge computes the base64 solution for a given Argon2 base64 
// challenge within deadlineUnixMilli milliseconds, if any was found. Deadlines are measured
// on the wall clock, not the monotonic clock, due to unreliability on mobile devices.
func Argon2PreimageChallenge(b64Challenge string, deadlineUnixMilli uint64) (b64Solution string, err error) {
```

**Review by wussler — APPROVED** [review]

**Pull request merged**